### PR TITLE
Catch errors in unicorn

### PR DESCRIPTION
**Pull request by antmat**
Catch exceptions in unicorn calls and process calbacks in executor to prevent deadlocks.

### CODE REVIEW

File: unicorn/src/unicorn/zookeeper.cpp [hunk 1]
```diff
@@ -96,27 +97,42 @@ zookeeper_t::put(
...
         path,
         value,
         std::move(version)
     );
+
     COCAINE_LOG_DEBUG(logger, "start processing");
-    zk.put(handler.path, handler.encoded_value, handler.version, handler);
+    try {
+        zk.put(handler.path, handler.encoded_value, handler.version, handler);
+    } catch (...) {
```
Comment (3Hren) [inline unicorn/src/unicorn/zookeeper.cpp]: Looks like boilerplate, isn't it?

**Review by 3Hren — CHANGES_REQUESTED** [review]

File: unicorn/src/unicorn/zookeeper.cpp (hunk 1 above)
Comment (antmat) [reply unicorn/src/unicorn/zookeeper.cpp]: suggestions?

Comment (3Hren) [reply unicorn/src/unicorn/zookeeper.cpp]: goto

Comment (antmat) [reply unicorn/src/unicorn/zookeeper.cpp]: nice catch! I think I can use it

Comment (antmat) [reply unicorn/src/unicorn/zookeeper.cpp]: try_run([]{
        zk.put(handler.path, handler.encoded_value, handler.version, handler);
    }, handler.callback, *executor);

Would be better?

**Comment by antmat** [comment]
@3Hren @karitra PTAL

**Review by karitra — APPROVED** [review]
LGTM

File: unicorn/src/unicorn/zookeeper.cpp [hunk 2]
```diff
@@ -158,16 +186,23 @@ zookeeper_t::create(callback::create callback, const path_t& path, const value_t
...
 scope_ptr
 zookeeper_t::del(callback::del callback, const path_t& path, version_t version) {
     auto handler = std::make_unique<del_action_t>(callback);
-    zk.del(path, version, std::move(handler));
-    return nullptr;
+    try_run([&]{
+        zk.del(path, version, std::move(handler));
+    }, callback, *executor);
+
+    auto scope = std::make_shared<zk_scope_t>();
```
Comment (3Hren) [inline unicorn/src/unicorn/zookeeper.cpp]: What's the purpose of this variable?

File: unicorn/src/unicorn/zookeeper.cpp [hunk 3]
```diff
@@ -201,32 +238,38 @@ zookeeper_t::children_subscribe(callback::children_subscribe callback, const pat
...
+    try_run([&]{
+        zk.childs(handler.path, handler, handler);
+    }, handler.callback, *executor);
     COCAINE_LOG_DEBUG(logger, "enqueued");
     return scope;
 }
 
 scope_ptr
 zookeeper_t::increment(callback::increment callback, const path_t& path, const value_t& value) {
+    auto scope = std::make_shared<zk_scope_t>();
```
Comment (3Hren) [inline unicorn/src/unicorn/zookeeper.cpp:247]: And this?

File: unicorn/src/zookeeper/connection.cpp [hunk 4]
```diff
@@ -248,9 +242,10 @@ connection_t::handle_ptr connection_t::init() {
 }
 
 void connection_t::close(zhandle_t* handle) {
-    io_loop.post([=]{
+    std::function<void()> f = [=]{
```
Comment (3Hren) [inline unicorn/src/zookeeper/connection.cpp]: And this?

**Review by 3Hren — CHANGES_REQUESTED** [review]

File: unicorn/src/unicorn/zookeeper.cpp [hunk 5]
```diff
@@ -42,6 +41,7 @@
 #include <memory>
 #include <blackhole/wrapper.hpp>
 
+
```
Comment (3Hren) [inline unicorn/src/unicorn/zookeeper.cpp]: Excess.

File: unicorn/src/unicorn/zookeeper.cpp [hunk 6]
```diff
@@ -75,11 +75,25 @@ public api::unicorn_scope_t
     close(){}
 };
 
+template<class F, class Result>
+auto try_run(F runnable, std::function<void(std::future<Result>)>& callback, api::executor_t& executor) -> void {
```
Comment (3Hren) [inline unicorn/src/unicorn/zookeeper.cpp:78]: Make static or put into anonymous namespace (preferred).

File: unicorn/src/unicorn/zookeeper.cpp [hunk 7]
```diff
@@ -89,34 +103,45 @@ zookeeper_t::zookeeper_t(cocaine::context_t& _context, const std::string& _name,
 
 zookeeper_t::~zookeeper_t() = default;
 
+
```
Comment (3Hren) [inline unicorn/src/unicorn/zookeeper.cpp]: Excess.

**Review by 3Hren — CHANGES_REQUESTED** [review]

**Comment by 3Hren** [comment]
And the biggest trouble: it doesn't compile.

File: unicorn/src/unicorn/zookeeper.cpp (hunk 3 above)
Comment (antmat) [reply unicorn/src/unicorn/zookeeper.cpp:247]: it's a scope for zk handlers.

File: unicorn/src/unicorn/zookeeper.cpp (hunk 6 above)
Comment (antmat) [reply unicorn/src/unicorn/zookeeper.cpp:78]: It's already there

**Comment by antmat** [comment]
considering build - looks like build started before PR in core was merged, because it builds fine on my linux dev server and mac.

**Comment by antmat** [comment]
and because core can not be built on linux)

**Pull request merged**